compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/ProvidesMethodFactoryCodeGen.kt

```diff
@@ -0,0 +1,541 @@
+import com.squareup.kotlinpoet.FileSpec
+import com.squareup.kotlinpoet.FunSpec
+import com.squareup.kotlinpoet.KModifier.OVERRIDE
+import com.squareup.kotlinpoet.KModifier.PRIVATE
+import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
+import com.squareup.kotlinpoet.PropertySpec
+import com.squareup.kotlinpoet.TypeName
+import com.squareup.kotlinpoet.TypeSpec
+import com.squareup.kotlinpoet.asClassName
+import com.squareup.kotlinpoet.jvm.jvmStatic
+import com.squareup.kotlinpoet.ksp.toClassName
+import com.squareup.kotlinpoet.ksp.toTypeName
+import com.squareup.kotlinpoet.ksp.toTypeParameterResolver
+import com.squareup.kotlinpoet.ksp.writeTo
+import dagger.Provides
+import dagger.internal.Factory
+import dagger.internal.Preconditions
+import org.jetbrains.kotlin.descriptors.ModuleDescriptor
+import org.jetbrains.kotlin.psi.KtFile
+import java.io.File
+
+object ProvidesMethodFactoryCodeGen : AnvilApplicabilityChecker {
+  override fun isApplicable(context: AnvilContext) = context.generateFactories
+
+  internal class KspGenerator(
+    override val env: SymbolProcessorEnvironment,
+  ) : AnvilSymbolProcessor() {
+    @AutoService(SymbolProcessorProvider::class)
+    class Provider : AnvilSymbolProcessorProvider(ProvidesMethodFactoryCodeGen, ::KspGenerator)
+
+    override fun processChecked(resolver: Resolver): List<KSAnnotated> {
+      resolver.getSymbolsWithAnnotation(daggerModuleFqName.asString())
+        .filterIsInstance<KSClassDeclaration>()
+        .forEach { clazz ->
+          val classAndCompanion = sequenceOf(clazz)
+            .plus(
+              clazz.declarations.filterIsInstance<KSClassDeclaration>().filter { it.isCompanionObject },
+            )
+
+          val functions = classAndCompanion.flatMap { it.getDeclaredFunctions() }
+            .filter { it.isAnnotationPresent<Provides>() }
+            .onEach { function ->
+              checkFunctionIsNotAbstract(clazz, function)
+            }
+            .also { functions ->
+              assertNoDuplicateFunctions(clazz, functions)
+            }
+            .map { function ->
+              CallableReference.from(function)
+            }
+
+          val properties = classAndCompanion.flatMap { it.getDeclaredProperties() }
+            .filter { it.isAnnotationPresent<Provides>() || it.getter?.isAnnotationPresent<Provides>() == true }
+            .filter { property ->
+              // Must be '@get:Provides'.
+              (property.getKSAnnotationsByType(Provides::class).singleOrNull()?.useSiteTarget == GET) ||
+                property.getter?.isAnnotationPresent<Provides>() == true
+            }
+            .map { property ->
+              CallableReference.from(property)
+            }
+
+          val className = clazz.toClassName()
+          val containingFile = clazz.containingFile!!
+          // TODO we need a public API for this in KSP
+          //  https://github.com/google/ksp/issues/1621
```

Comment:
They're open to this, so I'll send a PR to them soon. In the meantime, we do this internal access in #713 anyway so it's not necessarily new